notion-core/src/main/scala/com/ambiata/notion/core/LineReducer.scala

```diff
@@ -0,0 +1,104 @@
+package com.ambiata.notion
+package core
+
+import java.security.MessageDigest
+
+import org.apache.commons.codec.binary.Base64
+
+import scala.collection.mutable.Queue
+import scalaz.{Foldable, Monoid}
+
+/**
+ * "Reducer-like" class to accumulate values while
+ * reading a file and transforming them to a final value
+ * when the whole file has been read
+ */
+trait LineReducer[T] {
+  type S
+  
+  /** initial value */
+  def init: S
+
+  /** reducer function using the current line and the current state S */
+  def reduce(line: String, s: S): S
+
+  /** final value from the last state */
+  def finalise(s: S): T
```

Comment:
I am a touch skeptical of having two type parameters here. This trick is to allow horizontal composition when you have to do extra computation on output - which is explicitly not the case here. Not a major deal - just seems like it isn't necessary and composition will be improved without.